File: tools/component-tool/src/component_tool/report_writer.py
"""Report writer - generates JSON and Markdown reports from review results."""

from __future__ import annotations

import json
from pathlib import Path

from .models import ReviewResult, ReviewRun


def write_reports(
    run: ReviewRun,
    output_dir: Path,
) -> Path:
    """Write all reports for a review run.

    Creates:
      <output_dir>/<run_id>/summary.json
      <output_dir>/<run_id>/summary.md
      <output_dir>/<run_id>/<component_id>.json
      <output_dir>/<run_id>/<component_id>.md
      <output_dir>/<run_id>/<component_id>.raw.txt

    Returns the run directory path.
    """
    run_dir = output_dir / run.run_id
    run_dir.mkdir(parents=True, exist_ok=True)

    # Write per-component reports
    for cid, result in run.results.items():
        _write_component_json(result, run_dir)
        _write_component_md(result, run_dir)
        _write_component_raw(result, run_dir)

    # Write summary
    _write_summary_json(run, run_dir)
    _write_summary_md(run, run_dir)

    return run_dir


def _safe_filename(component_id: str) -> str:
    """Convert a component_id to a safe filename."""
    return component_id.replace("/", "_").replace("\\", "_")


def _write_component_json(result: ReviewResult, run_dir: Path) -> None:
    """Write per-component JSON report."""
    filename = _safe_filename(result.component_id)
    path = run_dir / f"{filename}.json"
    data = result.model_dump(mode="json", exclude={"raw_agent_output"})
    path.write_text(json.dumps(data, indent=2) + "\n", encoding="utf-8")


def _write_component_md(result: ReviewResult, run_dir: Path) -> None:
    """Write per-component Markdown report."""
    filename = _safe_filename(result.component_id)
    path = run_dir / f"{filename}.md"

    lines = [
        f"# Review: {result.component_id}",
        "",
        f"**Status:** {result.status.value}",
        "",
    ]

    if result.issues:
        lines.append("## Issues")
        lines.append("")
        for issue in result.issues:
            lines.append(f"### {issue.issue_id}. [{issue.severity.value.upper()}] {issue.title}")
            lines.append("")
            lines.append(f"**Kind:** {issue.kind.value}")
            lines.append("")
            lines.append(issue.details)
            lines.append("")
            if issue.evidence:
                lines.append("**Evidence:**")
                for ev in issue.evidence:
                    loc = f"{ev.path}"
                    if ev.lines:
                        loc += f":{ev.lines}"
                    lines.append(f"- `{loc}`")
                lines.append("")
            if issue.suggested_action:
                lines.append(f"**Suggested action:** {issue.suggested_action}")
                lines.append("")
    else:
        lines.append("No issues found.")
        lines.append("")

    if result.proposed_invariants:
        lines.append("## Proposed Invariants")
        lines.append("")
        for inv in result.proposed_invariants:
            lines.append(
                f"- `{inv.source_component_id}` -> `{inv.target_component_id}`: "
                f"{inv.expectation}"
            )
        lines.append("")

    path.write_text("\n".join(lines), encoding="utf-8")


def _write_component_raw(result: ReviewResult, run_dir: Path) -> None:
    """Write raw agent output."""
    filename = _safe_filename(result.component_id)
    path = run_dir / f"{filename}.raw.txt"
    path.write_text(result.raw_agent_output or "(no raw output)", encoding="utf-8")


def _write_summary_json(run: ReviewRun, run_dir: Path) -> None:
    """Write summary JSON report."""
    path = run_dir / "summary.json"
    data = {
        "run_id": run.run_id,
        "components_reviewed": run.components_reviewed,
        "results": {},
    }
    for cid, result in run.results.items():
        data["results"][cid] = {
            "status": result.status.value,
            "issue_count": len(result.issues),
            "proposed_invariant_count": len(result.proposed_invariants),
        }
    path.write_text(json.dumps(data, indent=2) + "\n", encoding="utf-8")


def _write_summary_md(run: ReviewRun, run_dir: Path) -> None:
    """Write summary Markdown report."""
    path = run_dir / "summary.md"

    lines = [
        f"# Review Summary: {run.run_id}",
        "",
        f"**Components reviewed:** {len(run.components_reviewed)}",
        "",
        "## Results",
        "",
        "| Component | Status | Issues | Proposed Invariants |",
        "|-----------|--------|--------|---------------------|",
    ]

    for cid in run.components_reviewed:
        result = run.results.get(cid)
        if result:
            lines.append(
                f"| `{cid}` | {result.status.value} | "
                f"{len(result.issues)} | {len(result.proposed_invariants)} |"
            )

    lines.append("")
    path.write_text("\n".join(lines), encoding="utf-8")

File: tools/component-tool-ts/src/reportWriter.ts
/**
 * Report writer - generates JSON and Markdown reports from review results.
 */

import { existsSync, mkdirSync, writeFileSync } from "fs";
import { join } from "path";
import type { ReviewResult, ReviewRun } from "./models.js";

function safeFilename(componentId: string): string {
  return componentId.replace(/[/\\]/g, "_");
}

export function writeReports(run: ReviewRun, outputDir: string): string {
  const runDir = join(outputDir, run.run_id);
  mkdirSync(runDir, { recursive: true });

  for (const [cid, result] of Object.entries(run.results)) {
    writeComponentJson(result, runDir);
    writeComponentMd(result, runDir);
    writeComponentRaw(result, runDir);
  }

  writeSummaryJson(run, runDir);
  writeSummaryMd(run, runDir);

  return runDir;
}

function writeComponentJson(result: ReviewResult, runDir: string): void {
  const filename = safeFilename(result.component_id);
  const data = { ...result, raw_agent_output: undefined };
  delete (data as Record<string, unknown>).raw_agent_output;
  writeFileSync(
    join(runDir, `${filename}.json`),
    JSON.stringify(data, null, 2) + "\n",
    "utf-8",
  );
}

function writeComponentMd(result: ReviewResult, runDir: string): void {
  const filename = safeFilename(result.component_id);
  const lines: string[] = [
    `# Review: ${result.component_id}`,
    "",
    `**Status:** ${result.status}`,
    "",
  ];

  if (result.issues.length > 0) {
    lines.push("## Issues", "");
    for (const issue of result.issues) {
      lines.push(
        `### ${issue.issue_id}. [${issue.severity.toUpperCase()}] ${issue.title}`,
        "",
        `**Kind:** ${issue.kind}`,
        "",
        issue.details,
        "",
      );
      if (issue.evidence.length > 0) {
        lines.push("**Evidence:**");
        for (const ev of issue.evidence) {
          const loc = ev.lines ? `${ev.path}:${ev.lines}` : ev.path;
          lines.push(`- \`${loc}\``);
        }
        lines.push("");
      }
      if (issue.suggested_action) {
        lines.push(`**Suggested action:** ${issue.suggested_action}`, "");
      }
    }
  } else {
    lines.push("No issues found.", "");
  }

  if (result.proposed_invariants.length > 0) {
    lines.push("## Proposed Invariants", "");
    for (const inv of result.proposed_invariants) {
      lines.push(
        `- \`${inv.source_component_id}\` -> \`${inv.target_component_id}\`: ${inv.expectation}`,
      );
    }
    lines.push("");
  }

  writeFileSync(join(runDir, `${filename}.md`), lines.join("\n"), "utf-8");
}

function writeComponentRaw(result: ReviewResult, runDir: string): void {
  const filename = safeFilename(result.component_id);
  writeFileSync(
    join(runDir, `${filename}.raw.txt`),
    result.raw_agent_output || "(no raw output)",
    "utf-8",
  );
}

function writeSummaryJson(run: ReviewRun, runDir: string): void {
  const data: Record<string, unknown> = {
    run_id: run.run_id,
    components_reviewed: run.components_reviewed,
    results: {} as Record<string, unknown>,
  };
  for (const [cid, result] of Object.entries(run.results)) {
    (data.results as Record<string, unknown>)[cid] = {
      status: result.status,
      issue_count: result.issues.length,
      proposed_invariant_count: result.proposed_invariants.length,
    };
  }
  writeFileSync(
    join(runDir, "summary.json"),
    JSON.stringify(data, null, 2) + "\n",
    "utf-8",
  );
}

function writeSummaryMd(run: ReviewRun, runDir: string): void {
  const lines: string[] = [
    `# Review Summary: ${run.run_id}`,
    "",
    `**Components reviewed:** ${run.components_reviewed.length}`,
    "",
    "## Results",
    "",
    "| Component | Status | Issues | Proposed Invariants |",
    "|-----------|--------|--------|---------------------|",
  ];

  for (const cid of run.components_reviewed) {
    const result = run.results[cid];
    if (result) {
      lines.push(
        `| \`${cid}\` | ${result.status} | ${result.issues.length} | ${result.proposed_invariants.length} |`,
      );
    }
  }

  lines.push("");
  writeFileSync(join(runDir, "summary.md"), lines.join("\n"), "utf-8");
}
